teamcity/pool_resource.go

```diff
@@ -0,0 +1,305 @@
+func (r *poolResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// get values from plan
+	var plan models.PoolDataModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// verify values
+	if plan.Name.IsNull() {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("name"),
+			"Agent Pool name cannot be null",
+			"The Resource cannot create an Agent Pool since there is an invalid configuration value for the Agent Pool name.",
+		)
+		return
+	}
+
+	// Generate API request
+	var pool models.PoolJson
+	var size int64
+
+	pool.Name = plan.Name.ValueString()
+	if !plan.Size.IsNull() {
+		size = plan.Size.ValueInt64()
+		pool.Size = &size
+	}
+
+	// Create new agent pool
+	result, err := r.client.NewPool(pool)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating pool",
+			"Cannot create pool, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	// Populate computed attributes
+	plan.Id = types.Int64Value(int64(*(result.Id)))
+
+	// Set state
+	diags = resp.State.Set(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// reads a resource and sets latest terraform state
+func (r *poolResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// get current state
+	var state models.PoolDataModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// get refreshed pool
+	pool, err := r.client.GetPool(state.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddAttributeError(
+			path.Root("name"),
+			"Agent Pool not found",
+			"The Datasource cannot get an Agent Pool since there is no Agent Pool with the provided name.",
+		)
+		return
+	}
+
+	// refresh state to reflect resource not found
+	if pool == nil {
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	// overwrite with refreshed state
+	state = models.PoolDataModel{
+		Name: types.StringValue(string(pool.Name)),
+		Size: pool.GetSize(),
+		Id:   types.Int64Value(int64(*(pool.Id))),
+	}
+
+	// set state
+	diags = resp.State.Set(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// updates a resource and sets the latest updated terraform state
+func (r *poolResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// get values from plan
+	var plan models.PoolDataModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// get state
+	var state models.PoolDataModel
+	diags = req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var newName string
+	var newSize string
+
+	// verify plan values
+	if plan.Name.IsNull() {
```

Comment:
Here and in the pool_data_source.go#78 too - maybe we could drop this code, wdyt? I suppose `"name": schema.StringAttribute{ Required: true,` is enough